sentry-android-core/src/main/java/io/sentry/android/core/AndroidTraceTransactionListener.java

```diff
@@ -0,0 +1,149 @@
+    this.hub = Objects.requireNonNull(hub, "hub is required");
+    final String tracesFilesDirPath = hub.getOptions().getProfilingTracesDirPath();
+    if (tracesFilesDirPath == null || tracesFilesDirPath.isEmpty()) {
+      hub.getOptions()
+          .getLogger()
+          .log(SentryLevel.ERROR, "No profiling traces dir path is defined in options.");
+      return;
+    }
+    traceFilesDir = new File(tracesFilesDirPath);
+  }
+
+  @Override
+  @SuppressWarnings("FutureReturnValueIgnored")
+  public synchronized void onTransactionStart(@NotNull ITransaction transaction) {
+
+    // Debug.startMethodTracingSampling() is only available since Lollipop
+    if (Build.VERSION.SDK_INT < Build.VERSION_CODES.LOLLIPOP) return;
+
+    // If a transaction is currently being profiled, we ignore this call
+    if (activeTransaction != null) {
+      return;
+    }
+
+    traceFile =
+        traceFilesDir == null ? null : new File(traceFilesDir, UUID.randomUUID() + ".trace");
+
+    if (traceFile == null) {
+      hub.getOptions().getLogger().log(SentryLevel.DEBUG, "Could not create a trace file");
+      return;
+    }
+
+    if (traceFile.exists()) {
+      hub.getOptions()
+          .getLogger()
+          .log(SentryLevel.DEBUG, "Trace file already exists: %s", traceFile.getPath());
+      return;
+    }
+
+    long intervalMillis = hub.getOptions().getProfilingTracesIntervalMillis();
+    if (intervalMillis <= 0) {
+      hub.getOptions()
+          .getLogger()
+          .log(
+              SentryLevel.DEBUG,
+              "Profiling trace interval is set to %d milliseconds",
+              intervalMillis);
+      return;
+    }
+    activeTransaction = transaction;
+
+    // We stop the trace after 30 seconds, since such a long trace is very probably a trace
+    // that will never end due to an error
+    scheduledFinish =
+        hub.getOptions()
+            .getExecutorService()
+            .schedule(() -> lastProfilingData = onTransactionFinish(transaction), 30_000);
+
+    int intervalUs = (int) MILLISECONDS.toMicros(intervalMillis);
+    Debug.startMethodTracingSampling(traceFile.getPath(), BUFFER_SIZE_BYTES, intervalUs);
+  }
+
+  @Override
+  public synchronized @Nullable ProfilingTraceData onTransactionFinish(
+      @NotNull ITransaction transaction) {
+
+    // Profiling finished, but we check if we cached last profiling data due to a timeout
+    if (activeTransaction == null) {
+      ProfilingTraceData profilingData = lastProfilingData;
+      // If the cached last profiling data refers to the transaction that started it we return it
+      // back, otherwise we would simply lose it
+      if (profilingData != null
+          && profilingData.getTraceId().equals(transaction.getSpanContext().getTraceId())) {
+        return profilingData;
+      }
+      return null;
+    }
+
+    if (activeTransaction != transaction) {
+      return null;
+    }
+
+    Debug.stopMethodTracing();
```

Comment:
there is some overhead (and possible frame loss), but that's because of profiling, not i/o writing. Moving things to background threads wouldn't make any difference